File: MiniDB_Projects/Team_SuperSymmetry/src/minidb/buffer_pool.py
"""
Buffer pool (page cache).

Holds a bounded number of frames in memory. Implements:
  * fetch/pin and unpin with reference counting
  * dirty-page tracking
  * LRU eviction of unpinned frames
  * the Write-Ahead-Logging rule: before a dirty data page is flushed to
    disk, the log must be durable up to that page's LSN. This is enforced
    via an optional `log_flush` callback wired to the WAL manager.

Statistics (hits/misses/evictions) are tracked so the buffer pool's effect
can be demonstrated and benchmarked.
"""
from __future__ import annotations

from collections import OrderedDict
from typing import Callable, Dict, Optional

from .disk_manager import DiskManager, PageId
from .page import Page


class Frame:
    __slots__ = ("page", "pin_count", "dirty")

    def __init__(self, page: Page):
        self.page = page
        self.pin_count = 0
        self.dirty = False


class BufferPool:
    def __init__(
        self,
        disk: DiskManager,
        capacity: int = 64,
        log_flush: Optional[Callable[[int], None]] = None,
    ):
        self.disk = disk
        self.capacity = capacity
        self.log_flush = log_flush  # called with a page LSN before flush
        self.frames: "OrderedDict[PageId, Frame]" = OrderedDict()
        self.hits = 0
        self.misses = 0
        self.evictions = 0

    # ---- core ------------------------------------------------------------
    def fetch_page(self, pid: PageId) -> Page:
        """Return the (pinned) Page for pid. Caller must unpin when done."""
        if pid in self.frames:
            self.hits += 1
            frame = self.frames[pid]
            frame.pin_count += 1
            self.frames.move_to_end(pid)  # mark most-recently-used
            return frame.page
        self.misses += 1
        if len(self.frames) >= self.capacity:
            self._evict_one()
        page = self.disk.read_page(pid)
        frame = Frame(page)
        frame.pin_count = 1
        self.frames[pid] = frame
        return page

    def new_page(self, file_key: str):
        """Allocate a fresh page on disk and return (PageId, Page). The
        returned Page is initialized (proper header), pinned, and dirty."""
        pid = self.disk.allocate_page(file_key)
        if len(self.frames) >= self.capacity:
            self._evict_one()
        page = Page()  # constructor initializes an empty slotted page
        frame = Frame(page)
        frame.pin_count = 1
        frame.dirty = True
        self.frames[pid] = frame
        return pid, page

    def unpin_page(self, pid: PageId, dirty: bool):
        frame = self.frames.get(pid)
        if frame is None:
            return
        if dirty:
            frame.dirty = True
        if frame.pin_count > 0:
            frame.pin_count -= 1

    def flush_page(self, pid: PageId):
        frame = self.frames.get(pid)
        if frame is None or not frame.dirty:
            return
        if self.log_flush is not None:
            self.log_flush(frame.page.page_lsn)  # WAL rule
        self.disk.write_page(pid, frame.page)
        frame.dirty = False

    def flush_all(self):
        for pid in list(self.frames.keys()):
            self.flush_page(pid)
        self.disk.fsync_all()

    # ---- eviction --------------------------------------------------------
    def _evict_one(self):
        for pid, frame in list(self.frames.items()):  # LRU order
            if frame.pin_count == 0:
                if frame.dirty:
                    self.flush_page(pid)
                del self.frames[pid]
                self.evictions += 1
                return
        raise RuntimeError("buffer pool full: all pages pinned")

    def stats(self) -> Dict[str, int]:
        total = self.hits + self.misses
        return {
            "hits": self.hits,
            "misses": self.misses,
            "evictions": self.evictions,
            "hit_rate_pct": round(100 * self.hits / total, 1) if total else 0.0,
        }

File: Team_Orchestrators/src/storage/buffer_pool.cpp
#include "minidb/storage/buffer_pool.hpp"

#include <stdexcept>

namespace minidb {

BufferPool::BufferPool(PageManager& pm, size_t num_frames) : pm_(pm) {
  if (num_frames == 0) num_frames = 1;
  frames_.resize(num_frames);
  free_.reserve(num_frames);
  for (size_t i = 0; i < num_frames; ++i) free_.push_back(i);
}

void BufferPool::lru_remove(size_t idx) {
  auto it = lru_pos_.find(idx);
  if (it != lru_pos_.end()) {
    lru_.erase(it->second);
    lru_pos_.erase(it);
  }
}

void BufferPool::lru_add(size_t idx) {
  lru_.push_back(idx);  // most-recently-used at the back
  lru_pos_[idx] = std::prev(lru_.end());
}

size_t BufferPool::acquire_frame() {
  if (!free_.empty()) {
    size_t idx = free_.back();
    free_.pop_back();
    return idx;
  }
  if (lru_.empty()) throw std::runtime_error("BufferPool: all pages pinned");
  size_t idx = lru_.front();
  lru_.pop_front();
  lru_pos_.erase(idx);
  Frame& f = frames_[idx];
  if (f.dirty) {
    pm_.write_page(f.page_id, f.data.data());
    f.dirty = false;
  }
  table_.erase(f.page_id);
  return idx;
}

uint8_t* BufferPool::fetch_page(PageId id) {
  auto it = table_.find(id);
  if (it != table_.end()) {
    ++hits_;
    size_t idx = it->second;
    if (frames_[idx].pin == 0) lru_remove(idx);
    ++frames_[idx].pin;
    return frames_[idx].data.data();
  }
  ++misses_;
  size_t idx = acquire_frame();
  Frame& f = frames_[idx];
  pm_.read_page(id, f.data.data());
  f.page_id = id;
  f.pin = 1;
  f.dirty = false;
  table_[id] = idx;
  return f.data.data();
}

PageId BufferPool::new_page(uint8_t*& out) {
  PageId id = pm_.allocate_page();
  size_t idx = acquire_frame();
  Frame& f = frames_[idx];
  pm_.read_page(id, f.data.data());  // zeroed page
  f.page_id = id;
  f.pin = 1;
  f.dirty = true;
  table_[id] = idx;
  out = f.data.data();
  return id;
}

void BufferPool::unpin(PageId id, bool dirty) {
  auto it = table_.find(id);
  if (it == table_.end()) return;
  size_t idx = it->second;
  Frame& f = frames_[idx];
  if (dirty) f.dirty = true;
  if (f.pin > 0) --f.pin;
  if (f.pin == 0) lru_add(idx);
}

void BufferPool::flush_all() {
  for (auto& f : frames_) {
    if (f.page_id != kInvalidPageId && f.dirty) {
      pm_.write_page(f.page_id, f.data.data());
      f.dirty = false;
    }
  }
  pm_.sync();
}

}  // namespace minidb
